econml/_ortho_learner.py

```diff
@@ -0,0 +1,584 @@
+        If int, random_state is the seed used by the random number generator;
+        If :class:`~numpy.random.mtrand.RandomState` instance, random_state is the random number generator;
+        If None, the random number generator is the :class:`~numpy.random.mtrand.RandomState` instance used
+        by `np.random`.
+
+    Examples
+    --------
+
+    The example code below implements a very simple version of the double machine learning
+    method on top of the :py:class:`~econml._ortho_learner._OrthoLearner` class, for expository purposes.
+    For a more elaborate implementation of a Double Machine Learning child class of the class
+    :py:class:`~econml._ortho_learner._OrthoLearner` checkout :py:class:`~econml.dml.DMLCateEstimator`
+    and its child classes.
+
+    .. highlight:: python
+    .. code-block:: python
+
+        import numpy as np
+        from sklearn.linear_model import LinearRegression
+        from econml._ortho_learner import _OrthoLearner
+        class ModelNuisance:
+            def __init__(self, model_t, model_y):
+                self._model_t = model_t
+                self._model_y = model_y
+            def fit(self, Y, T, W=None):
+                self._model_t.fit(W, T)
+                self._model_y.fit(W, Y)
+                return self
+            def predict(self, Y, T, W=None):
+                return Y - self._model_y.predict(W), T - self._model_t.predict(W)
+        class ModelFinal:
+            def __init__(self):
+                return
+            def fit(self, Y, T, W=None, nuisances=None):
+                Y_res, T_res = nuisances
+                self.model = LinearRegression(fit_intercept=False).fit(T_res.reshape(-1, 1), Y_res)
+                return self
+            def predict(self, X=None):
+                return self.model.coef_[0]
+            def score(self, Y, T, W=None, nuisances=None):
+                Y_res, T_res = nuisances
+                return np.mean((Y_res - self.model.predict(T_res.reshape(-1, 1)))**2)
+        np.random.seed(123)
+        X = np.random.normal(size=(100, 3))
+        y = X[:, 0] + X[:, 1] + np.random.normal(0, 0.1, size=(100,))
+        est = _OrthoLearner(ModelNuisance(LinearRegression(), LinearRegression()),
+                            ModelFinal(),
+                            n_splits=2, discrete_treatment=False, random_state=None)
+        est.fit(y, X[:, 0], W=X[:, 1:])
+
+    >>> est.score_
+    0.007568302109999707
+    >>> est.const_marginal_effect()
+    1.0236499258047582
+    >>> est.effect()
+    array([1.02364993])
+    >>> est.effect(T0=0, T1=10)
+    array([12.34401722])
+    >>> est.score(y, X[:, 0], W=X[:, 1:])
+    0.00727995424098179
+    >>> est.model_final.model
+    LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
+         normalize=False)
+    >>> est.model_final.model.coef_
+    array([1.02364993])
+
+    The following example shows how to do double machine learning with discrete treatments, using
```

Comment:
I think given the centrality of these classes and how many of us might use them to base other classes, I would like to have an extensive documention that keeps all the contracts that this class satisfies. If someone needs to make a change to this class that breaks any of these examples, then they need to think more before they do it. So having these examples and requiring that they are valid, might be a good think. Of course this can also happen with unit tests, but I want it to be hard for someone to change the contract that I am describing in the docstring :)